README
NAME
    WWW::YouTube::Download - Very simple YouTube video download interface

SYNOPSIS
      use WWW::YouTube::Download;
  
      my $client = WWW::YouTube::Download->new;
      $client->download($video_id);
  
      my $video_url = $client->get_video_url($video_id);
      my $title     = $client->get_title($video_id);     # maybe encoded utf8 string.
      my $fmt       = $client->get_fmt($video_id);       # maybe highest quality.
      my $suffix    = $client->get_suffix($video_id);    # maybe highest quality file suffix

DESCRIPTION
    WWW::YouTube::Download is a download video from YouTube.

METHODS
    new()
          $client = WWW::YouTube::Download->new;

        Creates a WWW::YouTube::Download instance.

    download($video_id [, \%args])
          $client->download($video_id);
          $client->download($video_id, {
              fmt      => 37,
              filename => 'sample.mp4', # save file name
          });
          $client->download($video_id, {
              filename => '{title}.{suffix}', # maybe `video_title.mp4`
          });
          $client->download($video_id, {
              cb => \&callback,
          });

        Download the video file. The first parameter is passed to YouTube
        video url.

        Allowed arguments:

        "cb"
            Set a callback subroutine, SEE LWP::UserAgent ':content_cb' for
            details.

        "filename"
            Set the filename, possibly using placeholders to be filled with
            information gathered about the video.

            "filename" supported format placeholders:

              {video_id}
              {title}
              {user}
              {fmt}
              {suffix}
              {resolution}

            Output filename is set to "{video_id}.{suffix}" by default.

        "file_name"
            DEPRECATED alternative for "filename".

        "fmt"
            set the format to download. Defaults to the best video quality
            (inferred by the available resolutions).

    playback_url($video_id, [, \%args])
          $client->playback_url($video_id);
          $client->playback_url($video_id, { fmt => 37 });

        Return playback URL of the video. This is direct link to the movie
        file. Function supports only "fmt" option.

    prepare_download($video_id)
        Gather data about the video. A hash reference is returned, with the
        following keys:

        "fmt"
            the default, suggested format. It is inferred by selecting the
            alternative with the highest resolution.

        "fmt_list"
            the list of available formats, as an array reference.

        "suffix"
            the filename extension associated to the default format (see
            "fmt" above).

        "title"
            the title of the video

        "user"
            the YouTube user owning the video

        "video_id"
            the video identifier

        "video_url"
            the URL of the video associated to the default format (see "fmt"
            above).

        "video_url_map"
            an hash reference containing details about all available
            formats.

        The "video_url_map" has one key/value pair for each available
        format, where the key is the format identifier (can be used as "fmt"
        parameter for "download", for example) and the value is a hash
        reference with the following data:

        "fmt"
            the format specifier, that can be passed to "download"

        "resolution"
            the resolution as *width*x*height*

        "suffix"
            the suffix, providing a hint about the video format (e.g. webm,
            flv, ...)

        "url"
            the URL where the video can be found

    ua([$ua])
          $self->ua->agent();
          $self->ua($LWP_LIKE_OBJECT);

        Sets and gets LWP::UserAgent object.

    get_video_url($video_id)
    get_title($video_id)
    get_user($video_id)
    get_fmt($video_id)
    get_fmt_list($video_id)
    get_suffix($video_id)

AUTHOR
    xaicron <xaicron {@} cpan.org>

CONTRIBUTORS
    yusukebe

SEE ALSO
    WWW::NicoVideo::Download

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl module WWW::YouTube::Download

0.56 2013-07-30T07:14:40Z
    - Fixed typo: $Carp::Intrenal -> $Carp::Internal (Syaltut++)
    - Changes file reformatted as per CPAN::Changes::Spec (neilbowers++)

0.55 2013-07-27T09:21:38Z
    - Changed signature encoding (kucharskim++)

0.54 2013-07-05T18:29:31Z
    - If unavailable video, croak with an appropriate error message.
      (kucharskim++)
    - Support encrypted signature video. (kucharskim++)

0.53 2013-06-05T16:05:56Z
    - Various improvements to youtube-playlists (kucharskim++)

0.52 2013-05-05T16:18:54Z
    - Added playlist_id() and video_id() method (kucharskim++)

0.51 2013-05-03T07:49:29Z
    - Support new url (mikolaj++)

0.50 2013-04-19T17:50:39Z
    - repackaging

0.49 2013-04-09T21:13:56Z
    - repacking

0.48 2013-04-09T21:09:55Z
    - Fixed parsing problem on youtube-playlists (mikolaj++)

0.47 2013-04-02 02:06:07
    - Added some documentation and added eg/menu.pl (polettix++)

0.46 2013-03-24 15:34:27
    - YouTube's design was changed (vjp++, yusukebe++)

0.45 2013-02-10 02:14:09
    - New substitution variable {user} (isync++)

0.44 2013-01-19 18:29:58
    - added proxy support (isync++)

0.43 2013-01-05 01:43:17
    - fixed grammar (mokko++)
    - using Module::Build

0.42 2012-11-16 21:35:31
    - fixed "failed to extract JSON data" (ks0608++)
    - fixed RT: 81068 (panic83++)

0.41 2012-10-04 00:44:13
    - follow the change of site (mikolaj++, yusukebe++)
    - fixed color problems on scripts (mikolaj++)
    - fixed RT 79931 (HMBRAND++)

0.40 2012-01-22 16:54:53
    - Fixed a problem that did not works, because YouTube's design was changed.
      (ks0608++, zentooo++)

0.39 2012-01-18 01:28:00
    - remove suffix from youtube-download.pl and youtube-playlists.pl
      (mikolaj++)

0.38 2011-12-08 00:44:52
    - fix typo (mikolaj++)
    - remove wrong code (mikolaj++)

0.37 2011-12-07 02:33:23
    - fix wrong characters in filename in youtube-download.pl (mikolaj++)
    - fix doc in youtube-download.pl (mikolaj++)
    - file_name option is DEPRECATED

0.36 2011-10-31 02:45:48
    - fix default filename problem in youtube-download.pl

0.35 2011-10-18 00:28:58
    - fix typo in youtube-playlists.pl (rt:71721)

0.34 2011-10-15 16:14:57
    - workaround for win32
    - fix suffix problem

0.33 2011-10-12 01:17:34
    - added youtube-download.pl and youtube-playlist.pl (request by st3vil++)
    - fix doc

0.32 2011-08-23 01:52:20
    - added LICENSE file (no code change)

0.31 2011-08-23 00:25:02
    - added fmt 38 to .mp4 (mikolaj++)

0.30 2011-08-07 15:25:11
    - cleanup (mikolaj++)

0.29 2011-08-05 01:29:27
    - fixed a problem that did not works, becouse changed YouTube's design.
      (katsuhito konishi++)

0.28 2011-08-04 00:03:58
    - fixed a problem that did not works, becouse changed YouTube's design.
      (mikolaj++)

0.27 2011-05-29 23:32:31
    - fixed garbled (akiym++)
    - fixed typo (akiym++)

0.26 2011-04-20 01:34:55
    - fixed scrape problem (reported by Octavian++)

0.25 2011-03-19 00:39:41
    - added few more match URLs (mikolaj++)

0.24 2011-03-04 01:45:20
    - fixed could not fetch json data problem (akiym++)

0.23 2011-02-22 00:00:22
    - supported any video urls (mikolaj++)

0.22 2011-01-05 00:17:54
    - Fixed a problem that can not download videos there.
      (reported by Octavian)

0.21 2010-12-30 15:28:51
    - fixed bug for _video_id(). If already had $1 is defined,
      was returned $1 without matches. (woremacx++)

0.20 2010-12-15 00:57:43
    - YouTube's API had change, maybe fixed it.

0.16 2010-10-17 18:17:57
    - Required Text::More 0.96 later
    - fixed POD spell miss

0.15 2010-08-11 00:14:53
    - YouTube's API has changed so I fixed it. (reported by Bruce David)

0.14 2010-05-07 02:50:42
    - typo fix
    - wrote pod

0.13 2010-04-10 18:20:05
    - Each instance to create a LWP::UserAgent. # thx kuniyoshi

0.12 2010-04-03 16:06:33
    - upload miss X(

0.11 2010-04-03 15:02:25
    - added ua setter/getter.
      using CGI.

0.10 2010-04-03 02:56:18
    - rewrite all

0.07 2010-03-01 00:11:55
    - supported FULL HD quality

0.06 2009-11-08 00:46:55
    - Youtube respond to changes in
    
0.05 2009-10-12 14:58:23
    - Youtube respond to changes in

0.04 2011-08-14
    - Changes unknown

0.03 2011-08-12
    - Changes unknown

0.02 2009-06-06 21:04:17
    - added 'fmt' field

0.01 2009-04-04 16:26:29
    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    no_index    => { 'directory' => [ 'inc' ] },
    name        => 'WWW-YouTube-Download',
    module_name => 'WWW::YouTube::Download',

    script_files => [glob('script/*'), glob('bin/*')],

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/WWW/YouTube/Download.pm
package WWW::YouTube::Download;

use strict;
use warnings;
use 5.008001;

our $VERSION = '0.21';

use CGI ();
use Carp ();
use URI ();
use LWP::UserAgent;
use URI::Escape qw/uri_unescape/;
use JSON;

use constant DEFAULT_FMT => 18;

my $base_url = 'http://www.youtube.com/watch?v=';
my $info     = 'http://www.youtube.com/get_video_info?video_id=';

sub new {
    my $class = shift;
    my %args = @_;
    $args{ua} = LWP::UserAgent->new unless exists $args{ua};
    bless \%args, $class;
}

for my $name (qw[video_id video_url title fmt fmt_list suffix]) {
    no strict 'refs';
    *{"get_$name"} = sub {
        my $self = shift;
        my $video_id = shift || Carp::croak "Usage: $self->get_$name(\$video_id|\$watch_url)";

        my $data = $self->prepare_download($video_id);
        return $data->{$name};
    };
}

sub download {
    my $self = shift;
    my $video_id = shift || Carp::croak "Usage: $self->download('[video_id|video_url]')";
    my $args = shift || {};

    my $data = $self->prepare_download($video_id);

    my $fmt = $args->{fmt} || $data->{fmt} || DEFAULT_FMT;
    my $video_url = $data->{video_url_map}{$fmt}{url} || Carp::croak "this video has not supported fmt: $fmt";
    my $file_name = $args->{file_name} || $data->{video_id} . _suffix($fmt);

    $args->{cb} = $self->_default_cb({
        file_name => $file_name,
        verbose   => $args->{verbose},
    }) unless ref $args->{cb} eq 'CODE';

    my $res = $self->ua->get($video_url, ':content_cb' => $args->{cb});
    Carp::croak '!! $video_id download failed: ', $res->status_line if $res->is_error;
}

sub _default_cb {
    my $self = shift;
    my $args = shift;

    open my $wfh, '>', $args->{file_name} or die $args->{file_name}, " $!";
    binmode $wfh;
    return sub {
        my ($chunk, $res, $proto) = @_;
        print $wfh $chunk; # write file

        if ($self->{verbose} || $args->{verbose}) {
            my $size = tell $wfh;
            if (my $total = $res->header('Content-Length')) {
                printf "%d/%d (%f%%)\r", $size, $total, $size / $total * 100;
            }
            else {
                printf "%d/Unknown bytes\r", $size;
            }
        }
    };
}

sub prepare_download {
    my $self = shift;
    my $video_id = shift || Carp::croak "Usage: $self->prepare_download('[video_id|watch_url]')";
    $video_id = _video_id($video_id);

    return $self->{cache}{$video_id} if ref $self->{cache}{$video_id} eq 'HASH';

    local $Carp::CarpLevel = $Carp::CarpLevel + 1;

    my $params        = $self->_get_info($video_id);
    my $video_url_map = $self->_get_video_url_map($video_id);

    my $fmt_list = [];
    my $sorted = [
        map {
            push @$fmt_list, $_->[0]->{fmt};
            $_->[0]
        } sort {
            $b->[1] <=> $a->[1]
        } map {
            my $resolution = $_->{resolution};
            $resolution =~ s/(\d+)x(\d+)/$1 * $2/e;
            [ $_, $resolution ]
        } values %$video_url_map,
    ];

    my $hq_data = $sorted->[0];

    return $self->{cache}{$video_id} = {
        video_id      => $video_id,
        video_url     => $hq_data->{url},
        title         => $params->param('title'),
        video_url_map => $video_url_map,
        fmt           => $hq_data->{fmt},
        fmt_lsit      => $fmt_list,
        suffix        => $hq_data->{suffix},
    };
}

sub _get_info {
    my ($self, $video_id) = @_;

    local $Carp::CarpLevel = $Carp::CarpLevel + 1;

    my $url = "$info$video_id";
    my $res = $self->ua->get($url);
    Carp::croak "GET $url failed. status: ", $res->status_line if $res->is_error;

    my $params = CGI->new(uri_unescape $res->content);
    Carp::croak "$video_id not found" if $params->param('status') ne 'ok';

    return $params;
}

sub _get_video_url_map {
    my ($self, $video_id) = @_;

    local $Carp::CarpLevel = $Carp::CarpLevel + 1;

    my $args        = $self->_get_args($video_id);
    my $fmt_url_map = _parse_fmt_url_map($args->{fmt_url_map});
    my $fmt_map     = _parse_fmt_map($args->{fmt_map});

    my $video_url_map = +{
        map {
            $_->{fmt} => $_,
        } map +{
            fmt        => $_,
            resolution => $fmt_map->{$_},
            url        => $fmt_url_map->{$_},
            suffix     => _suffix($_),
        }, keys %$fmt_map
    };

    return $video_url_map;
}

sub _get_args {
    my ($self, $video_id) = @_;

    local $Carp::CarpLevel = $Carp::CarpLevel + 1;

    my $url = "$base_url$video_id";
    my $res = $self->ua->get($url);
    Carp::croak "GET $url failed. status: ", $res->status_line if $res->is_error;

    my $data;
    for (split "\n", $res->content) {
        if ($_ && /videoplayback/ && /signature/ && !/HTML/) {
            my ($json) = $_ =~ /^[^{]+(.*)[^}]+$/;
            $data = JSON->new->utf8(1)->decode($json);
            last;
        }
    }

    return $data->{args};
}

sub _parse_fmt_url_map {
    my $param = shift;
    my $fmt_url_map = {};
    for my $stuff (split ',', $param) {
        my ($fmt, $playback_url) = split '\|', $stuff, 2;
        $fmt_url_map->{$fmt} = $playback_url;
    }

    return $fmt_url_map;
}

sub _parse_fmt_map {
    my $param = shift;
    my $fmt_map = {};
    for my $stuff (split ',', $param) {
        my ($fmt, $resolution) = split '/', $stuff;
        $fmt_map->{$fmt} = $resolution;
    }

    return $fmt_map;
}

sub ua {
    my $self = shift;
    my $ua = shift || return $self->{ua};
    Carp::croak "Usage: $self->ua(\$LWP_LIKE_OBJECT)" unless eval { $ua->isa('LWP::UserAgent') };
    $self->{ua} = $ua;
}

sub _suffix {
    my $fmt = shift;
    return $fmt =~ /18|22|37/ ? '.mp4'
         : $fmt =~ /13|17/    ? '.3gp'
         :                      '.flv'
    ;
}

sub _video_id {
    my $stuff = shift;
    return $stuff =~ /watch\?v=([^&]+)/ ? $1 : $stuff;
}

1;
__END__

=head1 NAME

WWW::YouTube::Download - Very simply YouTube video download interface.

=head1 SYNOPSIS

  use WWW::YouTube::Download;
  
  my $client = WWW::YouTube::Download->new;
  $client->download($video_id);
  
  my $video_url = $client->get_video_url($video_id);
  my $title     = $client->get_title($video_id);     # maybe encoded utf8 string.
  my $fmt       = $client->get_fmt($video_id);       # maybe highest quality.

=head1 DESCRIPTION

WWW::YouTube::Download is a download video from YouTube.

=head1 METHODS

=over

=item B<new()>

  $client = WWW::YouTube::Download->new;

Creates a WWW::YouTube::Donwload instance.

=item B<download($video_id [, \%args])>

  $client->download($video_id);
  $client->download($video_id, {
      fmt       => 37,
      file_name => 'sample.mp4', # save file name
  });
  $client->download($video_id, {
      cb => \&callback,
  });

Download the video file.
The first parameter is passed to YouTube video url.
B<\&callback> details SEE ALSO L<LWP::UserAgent> ':content_cb'.

=item B<ua([$ua])>

  $self->ua->agent();
  $self->ua($LWP_LIKE_OBJECT);

Sets and gets LWP::UserAgent object.

=item B<get_video_url($video_id)>

=item B<get_title($video_id)>

=item B<get_fmt($video_id)>

=back

=head1 AUTHOR

Yuji Shimada

=head1 CONTRIBUTORS

yusukebe

=head1 SEE ALSO

L<WWW::NicoVideo::Download>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
